rust-runtime/aws-smithy-checksums/src/lib.rs

```diff
@@ -5,197 +5,251 @@
     }
-
-    fn make_new(&self) -> Box<dyn BodyCallback> {
-        Box::new(Crc32cCallback::default())
+    fn size(&self) -> u64 {
+        Self::size()
     }
 }
 
 #[derive(Debug, Default)]
-struct Sha1Callback {
+struct Sha1 {
     hasher: sha1::Sha1,
 }
 
-impl Sha1Callback {
+impl Sha1 {
     fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
         self.hasher.write_all(bytes)?;
-
         Ok(())
     }
 
-    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
-        let mut header_map = HeaderMap::new();
-        let key = HeaderName::from_static(SHA_1_NAME);
-        // We clone the hasher because `Hasher::finalize` consumes `self`
-        let hash = self.hasher.clone().finalize();
-        let value = HeaderValue::from_str(&base64::encode(&hash[..]))
-            .expect("base64 will always produce valid header values from checksums");
-
-        header_map.insert(key, value);
+    fn finalize(&self) -> Result<Bytes, BoxError> {
+        use sha1::Digest;
+        Ok(Bytes::copy_from_slice(
+            self.hasher.clone().finalize().as_slice(),
+        ))
+    }
 
-        Ok(Some(header_map))
+    // Size of the checksum in bytes
+    fn size() -> u64 {
+        20
     }
 }
 
-impl BodyCallback for Sha1Callback {
-    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
-        self.update(bytes)
+impl Checksum for Sha1 {
+    fn update(
+        &mut self,
+        bytes: &[u8],
+    ) -> Result<(), Box<(dyn std::error::Error + Send + Sync + 'static)>> {
+        Self::update(self, bytes)
     }
 
-    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
-        self.trailers()
+    fn finalize(&self) -> Result<Bytes, BoxError> {
+        Self::finalize(self)
     }
-
-    fn make_new(&self) -> Box<dyn BodyCallback> {
-        Box::new(Sha1Callback::default())
+    fn size(&self) -> u64 {
+        Self::size()
     }
 }
 
 #[derive(Debug, Default)]
-struct Sha256Callback {
+struct Sha256 {
     hasher: sha2::Sha256,
 }
 
-impl Sha256Callback {
+impl Sha256 {
     fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
         self.hasher.write_all(bytes)?;
-
         Ok(())
     }
 
-    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
-        let mut header_map = HeaderMap::new();
-        let key = HeaderName::from_static(SHA_256_NAME);
-        // We clone the hasher because `Hasher::finalize` consumes `self`
-        let hash = self.hasher.clone().finalize();
-        let value = HeaderValue::from_str(&base64::encode(&hash[..]))
-            .expect("base64 will always produce valid header values from checksums");
+    fn finalize(&self) -> Result<Bytes, BoxError> {
+        use sha2::Digest;
+        Ok(Bytes::copy_from_slice(
+            self.hasher.clone().finalize().as_slice(),
+        ))
+    }
 
-        header_map.insert(key, value);
+    // Size of the checksum in bytes
+    fn size() -> u64 {
+        32
```

Comment:
Not a big deal, but I think all the hash `Digest` types have a `output_size` function you can use for this value (e.g., https://docs.rs/sha2/0.10.2/sha2/trait.Digest.html#tymethod.output_size).

Reply:
TIL, thanks for pointing that out!